Fix maven dependency version parsing

nikita-tkachenko-datadog opened the pull request:
# What Does This Do
Fixes logic that parses Maven artifact versions to support versions that contain dash, e.g. `4.0.0-alpha-5`, `3.0.0-M6`

PerfectSlayer commented on dd-java-agent/instrumentation/maven-3.2.1/src/main/java/datadog/trace/instrumentation/maven3/MavenDependencyVersion.java:
```diff
@@ -1,19 +1,23 @@
   }
 
-  boolean isLaterThanOrEqualTo(MavenDependencyVersion other) {
+  public boolean isLaterThanOrEqualTo(MavenDependencyVersion other) {
```
What about implementing `Comparable<MavenDependencyVersion>`?

smola replied:
+1, also I think this still misses a few cases of Maven's version ordering rules: https://maven.apache.org/ref/3.5.2/maven-artifact/apidocs/org/apache/maven/artifact/versioning/ComparableVersion.html

PerfectSlayer commented on dd-java-agent/instrumentation/maven-3.2.1/src/main/java/datadog/trace/instrumentation/maven3/MavenDependencyVersion.java:
```diff
@@ -1,19 +1,23 @@
 
   private static final Pattern DOT = Pattern.compile("\\.");
   private final int[] tokens;
+  private final String remainder;
```
This is also called as _qualifier_ in Maven terminology.

PerfectSlayer commented on dd-java-agent/instrumentation/maven-3.2.1/src/main/java/datadog/trace/instrumentation/maven3/MavenDependencyVersion.java:
```diff
@@ -25,15 +29,35 @@ boolean isLaterThanOrEqualTo(MavenDependencyVersion other) {
+    if (tokens.length != other.tokens.length) {
+      return tokens.length >= other.tokens.length;
+    }
+    if (remainder == null && other.remainder != null) {
+      return true;
+    }
+    if (remainder != null && other.remainder == null) {
+      return false;
+    }
+    return Objects.compare(remainder, other.remainder, Comparator.naturalOrder()) > 0;
```
This can be simplified using `nullsFirst()` or `nullsLast()`: https://docs.oracle.com/javase/8/docs/api/java/util/Comparator.html#nullsFirst-java.util.Comparator-

PerfectSlayer reviewed (commented):
Some minor comments

nikita-tkachenko-datadog commented:
Oof, disregard all of this: I completely missed Maven's `ComparableVersion`. It takes care of many things and there's no point in replicating its logic given that it's available at runtime. I have removed my ad-hoc class and replaced its usages with the Maven's one.

PerfectSlayer reviewed (approved):
That’s a very valid way to fix it 😋
